test/integration/framework/etcd.go

```diff
@@ -182,22 +189,29 @@ func EtcdMain(tests func() int) {
 	// Bail out early when -help was given as parameter.
 	flag.Parse()
 
-	before := runtime.NumGoroutine()
-	stop, err := startEtcd()
+	before, stacktracesBefore := stacks(1 << 16)
+	stop, err := startEtcd(nil)
 	if err != nil {
 		klog.Fatalf("cannot run integration tests: unable to start etcd: %v", err)
 	}
 	result := tests()
 	stop() // Don't defer this. See os.Exit documentation.
 
+	if err := checkGoroutines(before, stacktracesBefore); err != nil {
+		klog.Fatalf("EtcdMain: %v", err)
+	}
+	os.Exit(result)
+}
+
+func checkGoroutines(before int, stacktracesBefore string) error {
 	checkNumberOfGoroutines := func() (bool, error) {
 		// We leave some room for leaked goroutines as there are
-		// still some leaks, mostly:
+		// still some leaks, mostly (?):
 		// - leak from lumberjack package we're vendoring
 		// - leak from apiserve healthz
 		// - leak from opencensus library
 		// Once fixed, we should be able to bring it down to zero.
-		if dg := runtime.NumGoroutine() - before; dg <= 3 {
+		if dg := runtime.NumGoroutine() - before; dg <= 2 {
```

Comment:
I think we should still change to 3 until you solve this one `leak from lumberjack package we're vendoring`.

Reply:
But it's passing with 2, right?

The problem with having this number too high was that a single leaked goroutine inside a single test didn't trigger the warning, but they still accumulate.

I have been thinking about this and it is probably better to decouple starting etcd and goroutine leak checking.

The end goal is to check after each test, with zero tolerance. The intermediate solution is to check at the end of the entire test run, with 2 or 3 as delta - then it doesn't matter which one we pick, because a reproducible leak in  each test will add up enough.

Let me make that change.

/hold

Reply:
> But it's passing with 2, right?

perf-test? yes, but failed with other tests, like `test/integration/apiserver/admissionwebhook`

Reply:
Good point, I hadn't looked at the test results yet.

That's another argument for factoring out the leak checking: how may leaked goroutines are acceptable should be determined by the test and not be hard-coded inside the checker.

Reply:
Or simply using https://github.com/uber-go/goleak?

Reply:
no please, keep this as it is, we don't want to regress, the number of leaked goroutines should be 0 always, we have this compromise now

Reply:
if we leave this a choice we'll be back to the thousands of goroutine leaked we had one year ago https://github.com/kubernetes/kubernetes/issues/108483 , it took @wojtek-t several months to reduce this to only 2

Reply:
oh, I misread, you are solving one and reducing the number, sorry, Thanks for doing that +1

Reply:
I was thinking we were going to allow one more 😅 ,

Reply:
> The end goal is to check after each test, with zero tolerance. The intermediate solution is to check at the end of the entire test run, with 2 or 3 as delta - then it doesn't matter which one we pick, because a reproducible leak in each test will add up enough.
> 
> Let me make that change.
> 
> /hold

/hold cancel

go go go